pkg/planner/core/logical_plans.go

```diff
@@ -1814,7 +1817,42 @@ func (ds *DataSource) fillIndexPath(path *util.AccessPath, conds []expression.Ex
 		}
 	}
 	err := detachCondAndBuildRangeForPath(ds.SCtx(), path, conds, ds.tableStats.HistColl)
-	return err
+	if err != nil {
+		return false, err
+	}
+
+	// check whether the corresponding is NULL under current conditions.
+	if len(path.Index.NoNullIdxColOffsets) > 0 {
+		includeNullRange := false
+
+		for _, col := range path.Index.NoNullIdxColOffsets {
+			for _, ran := range path.Ranges {
+				if len(ran.LowVal) <= col || len(ran.HighVal) <= col ||
+					ran.LowVal[col].IsNull() || ran.HighVal[col].IsNull() {
+					includeNullRange = true
+					break
+				}
+			}
+
+			if includeNullRange {
+				// If the condition can make sure the column is not null, we can also use this index.
+				// TODO: maybe only considering the conditions except accessConds is enough, because the ranges
+				// don't contain NULL.
+				if ranger.CheckColumnIsNotNullWithCNFConditions(ds.SCtx(), path.FullIdxCols[col], ds.allConds) {
```

Comment:
Indeed, `ranger.CheckColumnIsNotNullWithCNFConditions` can cover more cases than the previous "range" based assertion. However, the "range" based assertion is faster, so I kept them as a fast path.